Gstf extend

anilthanki opened the pull request.

Commit:
checks for protein codeing mRNAs while looking at longest CDS

Commit:
fixes Flake8 warnings

Commit:
removes debuging code

Commit:
extend FASTA header option to have gene symbol as header for easy comparison

Commit:
fixes travis errors

nsoranzo commented on tools/gstf_preparation/gstf_preparation.xml:
```diff
@@ -37,7 +35,11 @@ python '$__tool_directory__/gstf_preparation.py'
+
+        <param name="headers" type="select" display="radio" label="Change the header line of the FASTA sequences to the &gt;TranscriptId_species or GeneSymbol_species format" help="As required by TreeBest, part of the GeneSeqToFamily workflow">
+            <option value="transcript" selected="true">TranscriptId_species</option>
+            <option value="gene">GeneSymbol_species</option>
```
```suggestion
            <option value="gene">GeneSymbol_species</option>
            <option value="">Don't change</option>
```

nsoranzo commented on tools/gstf_preparation/gstf_preparation.xml:
```diff
@@ -13,9 +13,7 @@ python '$__tool_directory__/gstf_preparation.py'
 #for $fasta_input in $fasta_inputs
```
Increase the tool version

nsoranzo commented on tools/gstf_preparation/gstf_preparation.xml:
```diff
@@ -13,9 +13,7 @@ python '$__tool_directory__/gstf_preparation.py'
-#if $headers
-    --headers
-#end if
+--headers $headers
```
```suggestion
#if $headers
    --headers $headers
#end if
```

Commit:
adds Transcript ID in all labels, and adds an optio to add Transcipt symbol in FASTA header

nsoranzo commented on tools/gstf_preparation/gstf_preparation.xml:
```diff
@@ -37,7 +35,12 @@ python '$__tool_directory__/gstf_preparation.py'
+
+        <param name="headers" type="select" display="radio" label="Change the header line of the FASTA sequences to the &gt;TranscriptId_species or GeneSymbol_species format" help="As required by TreeBest, part of the GeneSeqToFamily workflow, only TranscriptId_species is acceptable format by Aequatus visualisation">
+            <option value="transcript" selected="true">TranscriptId_species</option>
+            <option value="gene">GeneSymbol_TrascnriptID_species</option>
```
```suggestion
            <option value="gene">GeneSymbol_TranscriptID_species</option>
```

nsoranzo commented on tools/gstf_preparation/gstf_preparation.xml:
```diff
@@ -37,7 +35,12 @@ python '$__tool_directory__/gstf_preparation.py'
+        <param name="headers" type="select" display="radio" label="Change the header line of the FASTA sequences to the &gt;TranscriptId_species or GeneSymbol_species format" help="As required by TreeBest, part of the GeneSeqToFamily workflow, only TranscriptId_species is acceptable format by Aequatus visualisation">
+            <option value="transcript" selected="true">TranscriptId_species</option>
+            <option value="gene">GeneSymbol_TrascnriptID_species</option>
+            <option value="transcript_symbol">TranscriptSymbol_TrascnriptID_species</option>
```
```suggestion
            <option value="transcript_symbol">TranscriptSymbol_TranscriptID_species</option>
```

nsoranzo replied:
Maybe add an option to leave the headers untouched (with empty value)?

Commit:
Update tools/gstf_preparation/gstf_preparation.xml

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

Commit:
Update tools/gstf_preparation/gstf_preparation.xml

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

Commit:
Update tools/gstf_preparation/gstf_preparation.xml

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

Commit:
fixes suggestions from git

Commit:
adds biotype and confidence in gene and representative filter in transcript

nsoranzo commented on tools/gstf_preparation/gstf_preparation.py:
```diff
@@ -122,6 +126,12 @@ def feature_to_dict(cols, parent_dict=None):
     global gene_count
     gene = feature_to_dict(cols)
+    if 'biotype' in gene and gene['biotype'] != 'protein_coding':
+        return
```
This filtering is not correct, there are genes with protein-coding transcripts which have a different biotype, see e.g. this gene from our test data:
https://www.ensembl.org/Mus_pahari/Gene/Summary?db=core;g=MGP_PahariEiJ_G0008413;r=13:62596741-62686932
It's classified as "polymorphic_pseudogene" but has 2 protein-coding transcripts.

I'm removing it.

nsoranzo commented on tools/gstf_preparation/gstf_preparation.py:
```diff
@@ -66,8 +66,10 @@ def create_tables(conn):
+        transcript_symbol VARCHAR,
         protein_id VARCHAR UNIQUE,
         protein_sequence VARCHAR,
+        biotype VARCHAR,
```
This seems redundant since we are already filtering out non-protein_coding transcripts ( https://github.com/TGAC/earlham-galaxytools/pull/123, for GFF3 files, see comment below).

anilthanki replied:
Sometimes Biotype mentioned in Gene line rather than transcript line so to avoid that posibility

nsoranzo replied:
I mean that's always going to be either `protein_coding` or NULL.

nsoranzo replied:
Unless you want to keep also the non-protein coding genes and transcripts in the database, and only do the filtering when writing the (merged and filtered) output FASTA file.

Let me know!

nsoranzo replied:
Decided to keep also the non-protein coding genes and transcripts in the database.

nsoranzo commented on tools/gstf_preparation/gstf_preparation.py:
```diff
@@ -429,10 +503,20 @@ def __main__():
+                    # Change the FASTA header to '>GeneSymbol_species', as required by TreeBest
+                    # Remove any underscore in the species
+                    gene_name = fetch_gene_symbol_for_gene(conn, fetch_gene_id_for_transcript(conn, transcript_id))
+                    entry.header = ">%s-%s_%s" % (gene_name, transcript_id, species_for_transcript.replace('_', ''))
```
Do you want GeneSymbol-TranscriptId_species (dash & underscore, as here) or GeneSymbol_TranscriptId_species (underscore & underscore, as in the select in the XML below)?

anilthanki replied:
Yes GeneSymbol_TranscriptId_species will break TreeBeST

nsoranzo commented on tools/gstf_preparation/gstf_preparation.py:
```diff
@@ -139,10 +149,15 @@ def add_transcript_to_dict(cols, species, transcript_dict):
     transcript = feature_to_dict(cols)
     if 'biotype' in transcript and transcript['biotype'] != 'protein_coding':
         return
```
At the moment this filtering of gene and transcripts is applied only to GFF3 files, not gene features in JSON format. Do we want to extend it to JSON files too?

anilthanki replied:
Yes.. I will do that...

nsoranzo replied:
Done.

nsoranzo committed:
Raise exceptions to log discarded lines

nsoranzo committed:
Fix `headers` param in tests

nsoranzo committed:
Make only one SELECT query per transcript

- Increase database meta version due to the previous addition of columns
to the `gene` table.
- More detailed select option values

nsoranzo commented on tools/gstf_preparation/gstf_preparation.py:
```diff
@@ -122,27 +126,33 @@ def feature_to_dict(cols, parent_dict=None):
     gene = feature_to_dict(cols)
+    if not gene['id']:
+        raise Exception("Id not found among column 9 attribute tags: %s" % cols[8])
+    if 'confidence' in gene and gene['confidence'] != 'high':
```
Is `confidence` an attribute found in Ensembl GFF3 files? Never saw it before (same for `representative` for transcripts).

nsoranzo committed:
Add select for canonical/coding/all transcript filtering

replacing the longest CDS boolean parameter.

- Filter transcript only for the FASTA output, not for the database.
- Add `biotype` column to `gene` table.
- Add `is_canonical` column to `transcript` table (found in Ensembl JSON
  files, equivalent to `representative` for EI in-house annotation).
- Reduce the number of SELECT query for efficiency.
- Move the check on `confidence` to `write_gene_dict_to_db()`, so it's
  applied also on JSON input.
- Update output test data accordingly.
- Add python 3.7 requirement.

nsoranzo commented on tools/gstf_preparation/gstf_preparation.py:
```diff
@@ -242,32 +246,51 @@ def write_gene_dict_to_db(conn, gene_dict):
         if gene is None:
             # This can happen when loading a JSON file from Ensembl
             continue
+        if 'confidence' in gene and gene['confidence'] != 'high':
```
Is `confidence` an attribute found in Ensembl GFF3 files? Never saw it before (same for `representative` for transcripts).

nsoranzo commented:
Ready from my side, please have a look!

anilthanki commented:
Had a quick look seems fine
